bridge/bridge.go

```diff
@@ -113,15 +119,48 @@ func (b *bridge) Close() error {
 
 func (b *bridge) SetValidators(validators []types.Address, threshold uint64) {
 	b.resetIsValidatorMap(validators)
+	b.validatorThreshold = threshold
+
 	b.sampool.UpdateValidatorSet(validators, threshold)
 }
 
-func (b *bridge) GetReadyMessages() []sam.ReadyMessage {
-	return b.sampool.GetReadyMessages()
+func (b *bridge) GetReadyMessages() ([]MessageWithSignatures, error) {
+	readyMessages := b.sampool.GetReadyMessages()
+
+	data := make([]MessageWithSignatures, 0, len(readyMessages))
+
+	for _, readyMsg := range readyMessages {
+		msg, ok := readyMsg.Data.(*Message)
+		if !ok {
+			return nil, fmt.Errorf("get unknown type data %T when fetching ready messages", readyMsg.Data)
+		}
+
+		data = append(data, MessageWithSignatures{
+			Message:    *msg,
+			Signatures: readyMsg.Signatures,
+		})
+	}
+
+	return data, nil
+}
+
+func (b *bridge) ValidateTx(tx *types.Transaction) error {
+	hash := getTransactionHash(tx)
+
+	if !b.sampool.IsMessageKnown(hash) {
+		return fmt.Errorf("unknown state transaction, hash=%s", hash.String())
+	}
```

Comment:
This is a potential problem.

You verify block state transactions by making sure that they're in your SAM pool - when this doesn't have to be the case. 

All that matters is that the signatures are ok. If by any chance you didn't have this txn in your SAM pool, then that means you'll fail this validation check and break out of the consensus cycle

Reply:
Totally agree with that. On validating we don't need to check SAM Pool. I'll fix that

Reply:
@zivkovicmilos You're right.
Fixed at d48cff7feefc0c0f7db635964e4bc6a609ee5e1f

Reply:
@zivkovicmilos 
If the block producer doesn't have enough signatures, then such transaction will not be added to block. If validator doesn't, won't send prepare message. But looks fine.
As you said, only approach to avoid that is seal signatures into transaction. It resolve bulk sync issue also

Reply:
As we discussed internally, this functionality should be part of a different PR 👍